Add parse_syntactic_spec

connor-a-francis opened the pull request:
Thank you for your help! Please read the following before contributing to
this project.

## Legal

This project and its contents are licensed under GPL-3.0 or greater.
See the LICENSE file at the root of this project. Your contributions must
therefore also be licensed under GPL-3.0 or greater.

Also by contributing to this project, you are signing off on the
[Developer Certificate of Origin (DCO)](https://developercertificate.org/)
asserting that your contributions may legally be licensed under GPL-3.0
or greater. If you do not want to sign off on the DCO, close or delete
this PR.

## Instructions

Make all the changes you want. When you are done, request a review.
Work with the maintainer to complete your contribution.
Complete the template below to your best ability.

---

# Template for Final Squash Commit Message

```
Make sure your subject line follows Conventional Commits.
Remember `feat:` trumps `fix:` and add an `!` if it contains a breaking change.
Delete any section that is not needed.

## BREAKING CHANGES

Non backwards compatible changes.

* #Issue-number - Blah blah blah

## Deprecations

Things that will disappear in future releases.

* #Issue-number - Blah blah blah

## Features

New functionality.
New functionality containing non-backwards compatible functionality are also listed above.

* #Issue-number - Blah blah blah

## Fixes

Bug fixes for existing functionality.
Bug fixes containing non-backwards compatible functionality are also listed above.

* #Issue-number - Blah blah blah

---

Those who made this possible:

Co-authored-by: First Last <email>
Co-authored-by: First Last <email>
Co-authored-by: First Last <email>
```

AksharP5 commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec.py:
```diff
@@ -0,0 +1,133 @@
+    if match is None:
+        raise MalformedLHSError(line)
+
+    non_terminal = match["non_terminal"]
+    alt_name = match["alt_name"]
+    return LhsNonTerminal(non_terminal, alt_name)
```
This can be simplified to improve readability and shorten the method:
```
return LhsNonTerminal(match["non_terminal"], match["alt_name"])
```

AksharP5 commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec.py:
```diff
@@ -0,0 +1,133 @@
+    return LhsNonTerminal(non_terminal, alt_name)
+
+
+def parse_right(line: Line, rhs: str) -> List[Symbol]:
+    total = []
+    if not rhs:
+        return total
+    symbols = rhs.split(" ")
+    for symbol in symbols:
+        if symbol == "":
+            continue
+        if symbol.startswith("#") or symbol is None:
+            return total
+        total.append(parse_symbol(symbol))
+    return total
```
This method is getting long, refactoring it would be helpful, I believe that this can be reduced to a couple lines while still handling all the logic.

AksharP5 commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec.py:
```diff
@@ -0,0 +1,133 @@
+def parse_symbol(symbol: str) -> Symbol:
+    capturing = re.match(r"<(?P<name>\S*)>(?P<alt_name>\S+)?", symbol)
+
+    if capturing:
+        name = capturing["name"]
+        alt_name = capturing["alt_name"]
+        return parse_capturing(name, alt_name)
+    return Terminal(symbol)
```
This can be compressed to shorten the function (Note that I am using a class-based approach in my refactor, along with camelCase): 
```
return self.parseCapturing(capturing["name"], capturing["alt_name"]) if capturing else Terminal(symbol)
```

AksharP5 commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec.py:
```diff
@@ -0,0 +1,133 @@
+
+
+def parse_capturing(name: str, alt_name: str) -> CapturingSymbol:
+    terminal = re.match(r"[A-Z_]+", name)
+
+    if alt_name is not None:
+        alt_name = alt_name.strip(":")
+
+    if terminal:
+        return CapturingTerminal(name, alt_name)
+    return RhsNonTerminal(name, alt_name)
```
This function can be compressed to reduce length: 
```
terminal = re.match(r"[A-Z_]+", name)
alt_name = alt_name.strip(":") if alt_name is not None else alt_name
return CapturingTerminal(name, alt_name) if terminal else RhsNonTerminal(name, alt_name)
```

AksharP5 commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec.py:
```diff
@@ -0,0 +1,133 @@
+    return empty is not None or line.string.startswith("#")
+
+
+def build_matches(
+    line: str,
+) -> tuple[Match[str] | None, Match[str] | None, Match[str] | None]:
+
+    standard = re.match(r"^\s*(?P<lhs><\S+>(?::\S+)?)\s*::=(?:\s(?P<rhs>.*))?", line)
+    separated = re.match(
+        r"^\s*(?P<lhs><\S+>(?::\S+)?)\s*\*\*=\s(?P<rhs>.*)(?P<separator>\+.*)",
+        line,
+    )
+    repeating = re.match(r"^\s*(?P<lhs><\S+>(?::\S+)?)\s*\*\*=\s(?P<rhs>.*)", line)
+
+    return standard, separated, repeating
```
Keeping the function definition on one line would improve readability in this case, even if it is longer than typical, and the separated variable can also be kept as one line, as it would make the code shorter and simpler to read, also the comma after the parameter type can be removed as there is only one parameter.

AksharP5 commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec.py:
```diff
@@ -0,0 +1,133 @@
+    return standard, separated, repeating
+
+
+def parse_syntactic_rule(line: Line) -> SyntacticRule:
+
+    standard, separated, repeating = build_matches(line.string)
+
+    if standard:
+        return StandardSyntacticRule(
+            line, parse_left(line, standard["lhs"]), parse_right(line, standard["rhs"])
+        )
+
+    elif separated:
+        return parse_repeating(
+            line,
+            separated["lhs"],
+            separated["rhs"],
+            separated["separator"].strip("+").strip(" "),
+        )
+
+    elif repeating:
+        return parse_repeating(line, repeating["lhs"], repeating["rhs"], None)
+
+    raise MalformedBNFError(line)
```
This method is fairly long, each statement can be split into its own helper function that can handle the logic.

AksharP5 commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec.py:
```diff
@@ -0,0 +1,133 @@
+def parse_repeating(
+    line: Line, left: str, right: str, separator: str | None
+) -> RepeatingSyntacticRule:
+    if separator:
+        return RepeatingSyntacticRule(
+            line,
+            parse_left(line, left),
+            parse_right(line, right),
+            Terminal(separator.strip("+")),
+        )
+    else:
+        return RepeatingSyntacticRule(
+            line, parse_left(line, left), parse_right(line, right), None
+        )
```
This can be compressed to improve readability, and reduce length:
```
return RepeatingSyntacticRule(line, parse_left, parse_right, Terminal(separator.strip("+")) if separator else None)
```

AksharP5 commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec.py:
```diff
@@ -0,0 +1,133 @@
+        )
+
+
+def parse_syntactic_spec(lines: List[Line | Divider] | None) -> SyntacticSpec:
+    s = SyntacticSpec()
+    if not lines:
+        return s
+    for line in lines:
+        if isinstance(line, Divider) or is_skippable(line):
+            continue
+        s.append(parse_syntactic_rule(line))
+    return s
```
I would recommend using a class-based approach and having this method outside of the class, so it can be called from outside, and have this method instantiate a helper class to handle the logic. Also, try to use camelCase for the function and variable names above, you can leave this function name as snake_case, and use _ before the names of the helper classes you define, below is a brief example of the class-based approach:
```
class SyntacticParserHelper
...
parser = SyntacticParserHelper(line)
```

AksharP5 commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec_test.py:
```diff
@@ -0,0 +1,353 @@
+    assert list(parse_syntactic_spec(lines)) == []
+
+
+def test_blank_rule():
+    lines = [makeDivider(), makeLine("<noun> ::=")]
+    expected = [
+        StandardSyntacticRule(makeLine("<noun> ::="), LhsNonTerminal("noun"), [])
+    ]
+    assert list(parse_syntactic_spec(lines)) == expected
```
Instances where you use the same line multiple times within the same test can be assigned to a variable, so that if anything needs to be changed, you will only need to change it in one place rather than everywhere, this can be applied to multiple tests below, also creating a method for creating objects such as the StandardSyntacticRule and LhsNonTerminal will be useful to not instantiate objects within the tests themselves:
```
line = makeLine("<noun> ::=")
```

AksharP5 reviewed (changes requested):
@ConnorFrancisWNE @StoneyJackson 

I have reviewed the changes and left some comments, please let me know if there are any questions, comments, or if something should be ignored. @StoneyJackson I have also went ahead and refactored the parse_syntactic_spec.py file on my own using a class-based approach while I was messing around with it, however I am not sure if you would like me to push these changes so that we can refactor further from that point, or if you want @ConnorFrancisWNE to do it himself?

AksharP5 commented:
After speaking with @ConnorFrancisWNE, we have decided to push my refactored changes to speed up the process, however he will still be refactoring the tests. @StoneyJackson please let us know if there is a need for a further refactor.

AksharP5 commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec_test.py:
```diff
@@ -0,0 +1,378 @@
+    blank_rule = makeLine("<noun> ::=")
+    lines = [makeDivider(), blank_rule]
+    expected = [
+        StandardSyntacticRule(blank_rule, LhsNonTerminal("noun"), [])
```
Splitting up the objects such as the Line and Divider into helper functions is great to reduce dependencies. I would recommend doing the same with these objects and any other objects that are seen below, this can be as simple as a helper function to create a LhsNonTerminal in this instance, the goal is to have it so if there is a change in how this is implemented, it will only need to be changed in one place as opposed to having to edit all of your tests currently. This can also be done with the syntactic rule, RhsNonTerminal, and Terminal objects. Other than that, I see no issues with your tests beyond this.

AksharP5 reviewed (changes requested):
@ConnorFrancisWNE 

I just have a slight recommendation regarding the tests, but after that I think you should be all set, and @StoneyJackson can provide you with further feedback. You may also want to squash your commits, co-author credentials can be provided in the section at the bottom of the commit, if you need a better understanding of how to do so, you can look at the previously merged PR for split rough spec, or I can explain further.

To Squash: 
```
git reset --soft reorg
git add .
git commit -m "your commit message, in this case it would be a feat:"
git push -f 
```

StoneyJackson commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec.py:
```diff
@@ -0,0 +1,137 @@
+        return self.spec
+
+
+class SyntacticParserHelper:
```
Let's improve the name...

This class is responsible for parsing a line in the syntactic section. So how about SyntacticLineParser?

StoneyJackson commented on src/plcc/parse_spec/parse_syntactic_spec/parse_syntactic_spec.py:
```diff
@@ -0,0 +1,137 @@
+    def _matchLeft(self) -> Match[str] | None:
+        return re.match(r"<(?P<nonTerminal>\S*)>(?::(?P<altName>\S+))?\s*", self.lhs)
+
+    def _isSkippable(self) -> bool:
```
_isSkippable is used beyond this class. So either restructure so it isn't or name it isSkippable (no underscore).

StoneyJackson replied:
This is only used in a method with a negation (a "not" in front). In such cases, I like to try to rename the method so that the negation isn't needed.

What about isSyntacticRule()?

You'll need to move the "not" into the method. But then that makes the call cleaner and more descriptive... sic:

    if not parser._isSkippable():
        self.spec.append(parser.parseSyntacticRule())
vs

    if parser.isSyntacticRule():
        self.spec.append(parser.parseSyntacticRule())

StoneyJackson reviewed (changes requested):
I think this is very close. Only a couple of minor changes.

connor-a-francis committed:
feat: parse syntactic spec

Parses a given syntactic spec and tokenizes the input into SyntacticRule
objects that contain Symbols on the left and right hand sides.

---

* Closes #6

---

Co-authored-by: Akshar Patel <aksharpatel1233@gmail.com>
